### ISSUE TITLE
Level-Dateien Layout

### ISSUE THREAD

**Issue by default0** [yoshis-fan/GoldenEggVine#1]
Levels haben einen Header, Objektbytes und Spritebytes.
Sollen alle Teile eines Levels jeweils in eine Datei geschrieben werden, sollen für header/objekt/spritebytes einzelne dateien angelegt werden, etc?

**Comment by yoshis-fan** [comment]
Der Level-Header ist Teil der Objektdaten, sollte deswegen auch zusammen mit ihnen abgespeichert werden; gleiches gilt für Screen-Exits. Der Aufbau der Dateien sollte so aussehen:

{Groß geschriebene Hex-ID}.obj
  Level-Header
  Objekt-Daten
  FF
  Screen-Exits
  FF

{Groß geschriebene Hex-ID}.spr
  Sprite-Daten
  FF FF

Die Dateien human-readable zu machen ist verschwendete Mühe, da sie vom Editor interpretiert und bearbeitet werden. Die Dateien selbst sollten Binary sein und kein Text.

Da die Entrances und Middle-Rings ebenfalls vom Editor bearbeitet werden, können diese auch ruhig Binary sein.
Per Default gibt es aber bei beiden diesen Daten Lücken, da die Translevel-Nummer auch jene Icons auf der Map einschließt, welche keine Level enthalten.
W1: 00-0B
W2: 0C-17
W3: 18-23
W4: 24-2F
W5: 30-3B
W6: 3C-47
![translevels](https://cloud.githubusercontent.com/assets/18013404/14044375/286ef1c6-f290-11e5-8d43-34516a992323.png)
Generell zeigen die unbenutzten Entrances auf dieselben Daten wie der Pointer für Translevel 00.
Die Pointer für Indizes 0A und 0B sind speziell und zeigen auf die Entrances der Intro-Szene und den Intro-Level.

Zum initialen Auslesen der Middle-Rings muss man allerdings etwas raffiniert an die Sache herangehen, um keine Duplikate zu bekommen. Details hierzu kann ich auf Anfrage gerne beitragen.

**Comment by default0** [comment]
.obj und .spr daten werden extrahiert. Siehe 84964dad75d8b422e221caaa244c8dc66bfef8f4

**Comment by default0** [comment]
Wo und in welchem format sollen midway und entrance-daten abgespeichert werden? Da beide abhängig von Trans-leveln (=overworld-map-icons?) sind, scheint es mir als sinnvoll diese Daten in ein separates Verzeichnis (overworld?/levelmenu?) zu extrahieren.

**Comment by yoshis-fan** [comment]
Meines Wissens nach gibt es bereits in der Project-Klasse Ordner, die dafür vorgesehen sind, namens "Entrances" und "Middle-Rings" oder so ähnlich. Falls nicht, bitte anlegen. Die Daten sollten gruppiert nach den Overworld-Levels werden. Dh in den Verzeichnissen liegen dann Dateien von 00 bis 47 (Hex). Die Entrances sind jeweils immer 4 Bytes groß, die Middle-Rings je nachdem, wie viele benutzt wurden. Also entweder 4 Bytes, 8 Bytes, 12 Bytes oder 16 Bytes. Der "Compiler" für die ROM sollte dann bei den Middle Rings nach der Dateigröße vorgehen, um zu bestimmen, wie viele für jedes Overworld-Level benutzt wurden (ist aber noch Zukunftsmusik).
Die Extrahierung der Middle Rings ist nicht ganz trivial, bei Problemen werde ich mich der Sache annehmen, bin zurzeit beschäftigt, sorry :/

**Comment by default0** [comment]
Translevel müssen immer mindestens einen Midring verwenden? Oder können midring-dateien auch leer sein wenn ein Translevel keine midrings verwendet?
Wie genau finde ich raus ob ich einen einzelnen midring datensatz ausgeben muss oder nicht?

**Comment by yoshis-fan** [comment]
Stimmt, sorry, können auch leer sein, z.B. 1-1 hat keinen Middle Ring, genauso wie Extra 1, Extra 2, Extra 3 und Extra 4 und evtl. noch ein paar andere, in dem Fall dann halt einfach eine leere Datei erzeugen.
Um Herauszufinden, wie viele Middle Rings in einem Level benutzt werden, musst du die Zuordnung Translevel => List&lt;Level&gt; aufstellen.
Das machst du folgendermaßen:
http://pastebin.com/FP1zj26N

Anschließend vergleichst du die 4 infrage kommenden Einträge in der Middle Ring Table mit den besuchten Levels. Alle, die dort nicht vorkommen, sind kein valider Middle Ring.
Diese Methode der rekursiven Level-Iteration von oben hat allerdings zwei Mäkel:
1 - Manche Levels haben Screen Exits, die nicht benutzt werden können, siehe Levels 0x36, 0x37 und 0x7D
2 - Es funktioniert nicht in Bowsers Schloss, wegen der 4 rotierenden Türen in Level 0x97

Die Destinations der 4 rotierenden Türen stehen an ROM-Offset 0x78B1F (4 Einträge á 4 Byte). Außerdem triggert das Einsetzen des Bowser-Kampfes Middle Ring №3 für Translevel 0x44. Ob das daran liegt, dass Level 0xDD Item Memory Index №3 hat, oder hardcoded ist, muss noch wissenschaftlich geprüft werden (ich gehe davon aus, dass Sprites 0x4F (Midring) und 0x8E (Bowser-Kampf) die beiden einzigen Sprites sind, die Middle Rings triggern).
Um diese Methode zu perfektionieren, müsste also auch gleichzeitig eine Überprüfung von Objekten (0x3C, 0xA5, 0xA6, 0xX6D, 0xX6E, 0xX6F, 0xX70, 0xX88) und Sprites (0x12, 0x42, 0x4E, 0x84, 0x93, 0xCA, 0xD0, 0xD1, 0x131, 0x147) stattfinden, welche ein Screen Exit triggern und außerdem müsste überprüft werden, ob ein Screen Exit von einem bestimmten Punkt im Level erreicht werden kann. (Listen evtl unvollständig...)

Alternativ kannst du auch die Überschneidungen zwischen den Pointern vergleichen, wobei hier die Gefahr besteht, unsauber zu arbeiten, wenn z.B. der erste benutzte Eintrag für einen Translevel der Middle Ring №1 ist; und nicht (wie es sein sollte), Index №0.

**Comment by default0** [comment]
pastebin ist ausgelaufen.

**Comment by default0** [comment]
Nach Implementierung der naiven Methode (rekursives suchen aller Screen Exits ohne erreichbarkeit oder ähnliches zu prüfen, ohne spezielles handling für bowser oder rotierende türen) haben die folgenden Trans-Levels keine midway daten: 0A, 0B, 2C, 08, 14, 20, 45, 47.

Ist das korrekt?

**Comment by yoshis-fan** [comment]
Ah, du lebst noch, Langeweile? =D
Verdammt, brauchst du das Pastebin-Gedöns noch?
Bin nicht 100%ig im Bilde, aber das sieht schonmal gut aus. Nach naivem Ansatz haben die meisten Translevels eh Midring-Daten (also auch leere Levels wie das Score-Icon und das Control-Icon, weil deren Entrances einfach blank sind und damit 0x0000 als Pointer haben).
Was auch fies ist, Nintendo hat die Pointer-Listen unvollständig gemacht, d.h. überflüssige Pointer (alles nach Extra 6) existieren nicht und überschneiden sich daher mit den eigentlichen Entrance/Midway-Daten.
Wäre nice, wenn man da irgendeinen Weg finden würde, das zu verbessern, z.B. detektieren, wenn eine ROM "frisch" ist und die leeren Indizes dann ignorieren. Aber ist Zukunftsmusik, weil man dafür auch noch überprüfen müsste, ob Secret-Levels exisitieren etc. pp, bzw dem User die Möglichkeiten geben, einen Hack mit 60 Levels zu erstellen oder noch andere Optionen für den Umfang des Projektes. Für den Moment sieht das aber gut aus.
In ein paar Wochen ist auch Uni vorbei, dann kann ich mich mal intensiver damit beschäftigen.

**Comment by default0** [comment]
Dann lass ich das mal vorerst so. Fehlen noch die Labels, die iirc auch aus der ROM extrahiert werden sollten - Infos dazu?

**Comment by yoshis-fan** [comment]
Keine Ahnung was du mit den Labels meinst...
Aber ich hab mal ein bisschen drüber nachgedacht, also eigentlich wäre es dann am besten, wenn man einfach nur eine Nutzer-Abfrage macht für die o.g. Pointer-Geschichte, also dass gefragt wird, ob "sinnlose" Pointer auch extrahiert werden sollen (und dann noch mit einem Tooltip o.Ä., dass man sagt, wenn es eine "frische" ROM ist, dann wird empfohlen, "sinnlose" Pointer außen vor zu lassen).
Aber ich muss mal gucken, heute Abend könnte ich auch wieder Zeit zum Skypen haben, dann können wir bisschen genauer quatschen.
